### PR TITLE
Lines and Dots

### PR DESCRIPTION
**Pull request by Alex-Tideman**
## Description & motivation 💭 
New UI for Visualizing and Exploring Workflows

**Goals:**

- Context - Combine visual with information. Inline all details of events. Give much clearer indication of failures with bold colors. Put pending activities inline instead of separated.
- Performance - Show a compact summary of events immediately, and then allow drilling into more details with timeline and history view. No need to paginate, scroll to view.
- Liveness - Poll on events and update visuals as new events come in. Refresh workflow on new events to get latest status and pending activities.
- Exploration - View child workflow timelines inline and zoom in/out to get a better big picture understanding. Compare multiple events at once. Understand latency between events.

The split of this feature begins at workflow-run-layout, that is where the LabsGuard is. From there all the files used should be independent of existing code. 

The only major change of existing UI is long-polling on events (removed auto-refresh toggle), removing the auto-refresh and adding the Relationships and Metadata tabs. There are also some css changes with colors that we'll wanna review.

## **Compact**

Linear representation of event history. Grouped by scheduled time, and then again grouped by activity type. Can expand groups to show all events. Clock-time representation of event history. Show durations of each event and the status. Can expand to show details and Timeline of child workflows.

Screen Shot 2024-03-20 at 2 05 46 PM
Screen Shot 2024-03-20 at 2 07 32 PM
Screen Shot 2024-03-20 at 2 06 23 PM

## **Timeline**

Clock-time representation of event history. Show durations of each event and the status. Can expand to show details and Timeline of child workflows.

Screen Shot 2024-03-20 at 2 05 55 PM
Screen Shot 2024-03-20 at 2 06 48 PM
Screen Shot 2024-03-20 at 2 07 03 PM

## **History**

Tree-style graph of event history with lines connecting related events. Can expand to show details and Timeline of child workflows.

Screen Shot 2024-03-20 at 2 06 04 PM
Screen Shot 2024-03-20 at 2 07 15 PM

### How was this tested 👻 

- [x] Manual testing
- [x] E2E tests added
- [x] Unit tests added

### Draft Checklist 
- [x] Optimize svg file size (use tool?) or virtualize rows, especially for History view.
- [x] Make timeline “live” waiting on pending (update ending line timestamp or recalculate durations or fetch pending activities?)
- [ ] More clear indication of concurrency (not just timestamp empty. Draw box or something around events with the same workflow task.
- [x] Make workflow tasks offset from main line and show connection (need to make workflow task groups)
- [x] Fix dots and lines on History view when filtering history or when new events come in
- [x] Fix Compact view and History view when opening drawer.
- [ ] Add Result icon or something for completed dot in timeline or icons to indicate payloads next to event name
- [x] Add Time axis labels. Labels should be relative except on the start/end lines (Timeline view only).
- [x] Show activity name on Local Activity name (should use ActivityType in details payload)
- [x] Fix failed child workflows (still shows as pending)
- [ ] Add ?id param to open page with detail drawer opened
- [x] Redesign the Strongly Typed errors to be inline
- [x] ADD UNIT TESTS TO THE MADNESS IN constants.ts
- [x] Continue to refactor and remove magic numbers for ratios
- [x] Figure out how to make text widths responsive
- [x] Flip dots to left and rows to right?
- [ ] Add second row per event to show more details on History View?
- [x] Integrate dark mode

### CODE REVIEW

File: src/lib/components/lines-and-dots/svg/group-details-row.svelte [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+  <Box point={[x, y]} {width} height={boxHeight} fill="#465A78" />
+  <Box point={[x, y]} {width} {height} fill="#1E293B" />
+  <Box
+    point={[x, y]}
+    width={gutter + 100}
+    {height}
+    fill={getStatusColor(status)}
+  />
+  <Text point={[x + gutter, y + 0.5 * height]} category="none">
+    {status}
```
Comment (laurakwhit) [inline src/lib/components/lines-and-dots/svg/group-details-row.svelte:112]: Looks like there's the possibility for this to be `undefined` ⬇️  (this is a Marker event)

Screenshot 2024-04-04 at 4 33 52 PM

File: src/lib/components/lines-and-dots/event-type-filter.svelte [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    }),
+  );
+
+  $: {
+    if (isVersionNewer('1.21', $temporalVersion)) {
+      options = options.filter(({ value }) => value !== 'update');
+    }
+  }
+
+  const onOptionClick = ({ value }) => {
```
Comment (laurakwhit) [inline src/lib/components/lines-and-dots/event-type-filter.svelte:30]: Seeing the following error when unselecting an event type with the event details open (in all three views) ⬇️ 
  
Screenshot 2024-04-04 at 5 02 46 PM

It appears to leave the page in a broken state.

File: src/lib/layouts/workflow-history-layout-v2.svelte [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+            data-testid="compact"
+            on:click={() => ($eventViewType = 'compact')}
+            >{translate('workflows.compact')}</ToggleButton
+          >
+          <ToggleButton
+            active={$eventViewType === 'timeline'}
+            data-testid="timeline"
+            on:click={() => ($eventViewType = 'timeline')}
+            >{translate('common.timeline')}</ToggleButton
+          >
```
Comment (laurakwhit) [inline src/lib/layouts/workflow-history-layout-v2.svelte:110]: NIT: Noticed that if `Timeline` is selected and then labs mode is turned off `History` does not show up as selected (although it does default to the History view).

Screenshot 2024-04-04 at 5 18 49 PM

File: src/lib/components/lines-and-dots/event-type-filter.svelte (hunk 2 above)
Comment (Alex-Tideman) [reply src/lib/components/lines-and-dots/event-type-filter.svelte:32]: Good catch, I now clear the active events when you pick a filter

File: src/lib/components/lines-and-dots/svg/group-details-row.svelte (hunk 1 above)
Comment (Alex-Tideman) [reply src/lib/components/lines-and-dots/svg/group-details-row.svelte]: 👍 I added a check to use group label if no status.

File: src/lib/layouts/workflow-history-layout-v2.svelte (hunk 3 above)
Comment (Alex-Tideman) [reply src/lib/layouts/workflow-history-layout-v2.svelte:110]: 👍 Fixed to show History as active when Timeline is the view but not on labs mode

**Pull request merged**